### ISSUE TITLE
Optimize ElasticSearch search queries to not always use Scroll API

### ISSUE THREAD

**Issue by porunov** [JanusGraph/janusgraph#1680]
We should optimize search queries to not always use Scroll API (for example on queries where only single entity is needed). Search API creates a context which is stored on ElasticSearch side (for some amount of time) and which is heavy (if created too many contexts). Some search queries do not require context to be opened at all but we open the context for any search query.

Related discussion: https://groups.google.com/forum/#!topic/janusgraph-dev/7BF9Qi7dbAs

**Comment by porunov** [comment]
Basically, we don't use queries for requests with setup `limit` which is less than `max-result-set-size`. I think we should note it somewhere in documentation that when you specify you limit less then `max-result-set-size` for ElasticSearch it will optimize your query to not use SearchAPI. Otherwise search API is used.

**Comment by porunov** [comment]
About usage of `search_after`. We could use this type of search but we need to have a unique sort field in the request as a last sort parameter. Sure, we could use `_id` field but for that we should make denormalization for vertices stored in ElasticSearch and store additional `vertex_id` id field in each elasticsearch document.

Another solution could be to indicate same fields as `unique`. Even so they will not be checked for uniqueness for mixed indexes, we could give a clue for the ElasticSearch that the field is considered to be unique (guaranteed by the application). If so, when this field is used as an any sort field for the query we can tell ElasticSearch that this field is tiebreaker for the query.

If the no tiebreaker fields is used than there are different possible scenarios which we could use:
1) We could include tiebreaker field with the smallest data type (i.e. `byte` if possible) as a last sort parameter for the query. (We should only check that there is no performance downgrade in this scenario).
2) We can include a tiebreaker field (copy of `_id` field) in all indices with no `unique` field. After that we can use this field if no `unique` field is specified in a search request.
3) Instead of option 1 or 2 (or both options) we can use Scroll API as a fallback. I.e. if either of options have some disadvantages, we always can fallback to use Scroll API.

**Comment by li-boxuan** [comment]
@porunov One question: why do you say "we can research a solution to imitate scroll usage by using search_after"? Is it because search_after is cheaper than scroll?

**Comment by li-boxuan** [comment]
Another approach is to use time-in-point API which ensures paging order. Not sure how is its performance compared to scrolling.

**Comment by porunov** [comment]
> @porunov One question: why do you say "we can research a solution to imitate scroll usage by using search_after"? Is it because search_after is cheaper than scroll?

search_after stores all the necessary filters + ordering state inside a special token. So, you don't create context in ElasticSearch at all.
Scroll API on the other hand create a scroll context which is stored inside ElasticSearch for a specified amount of time. Usually TTL for that context is updated each time you use it. After we reach TTL, the context is automatically removed from ElasticSearch.
Opening too many scrolls will result in degraded performance as far as I remember, thus the default maximum amount of opened scrolls in ElasticSearch is 500. That said, it can be configured to Integer.MAX_VALUE.
There is no such restrictions with search_after token because the search is state is not stored inside elasticsearch but instead stored inside `search_after` token.  
Comparing search performance of Scroll API and search_after is tricky and I assume will very on different things.
Probably, full scans or generally big scrolls will perform better with Scroll API and other search quries will perform better with search_after but it's just a random guess. I never actually compared the performance of Scroll API and search_after because banchmarks will very depending on different things (like amount of parallel opened scrolls, amount of filters / orders in search_after and Scroll API, cluster size, etc. etc.).

**Comment by cdegroc** [comment]
> Another approach is to use time-in-point API which ensures paging order. Not sure how is its performance compared to scrolling.

Agree. Starting with ES 7.10, the [PIT (point-in-time) API](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/point-in-time-api.html) would help avoid inconsistencies and would provide a tiebreaker field (`_shard_doc`) out of the box.

Quoting [the search_after documentation](https://www.elastic.co/guide/en/elasticsearch/reference/7.12/paginate-search-results.html#search-after):
> If you used a PIT, a tiebreaker is included as the last sort values for each hit. This tiebreaker called _shard_doc is added automically on every search requests that use a PIT. The _shard_doc value is the combination of the shard index within the PIT and the Lucene’s internal doc ID, **it is unique per document and constant within a PIT**. You can also add the tiebreaker explicitly in the search request to customize the order

The performance impact of using it, and compared to Scrolls, is unclear to me though.

**Comment by cdegroc** [comment]
Took a first stab at it using search_after + point-in-time.
I'd be happy to get feedback on the code design and whether it's useful as-is: it does not solve all issues but could be a first step towards using search_after?
